common/lib/common-utils/src/idleTaskScheduler.ts

```diff
@@ -0,0 +1,67 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryProperties } from "@fluidframework/common-definitions";
+
+
+interface Task {
+    handle: () => any,
+}
+
+
+/**
+ * Helper class that is used to schedule non-essential tasks
+ * Time measurements are in milliseconds as a floating point with a decimal
+ */
+export default class IdleTaskScheduler {
+    //List of tasks waiting to be run
+    private idleTaskList:  Task[] = [];
+    // A reference to the task currently being processed.
+    private taskHandle = 0;
+
+    /*
+    Add tasks to FIFO queue of tasks that are run during idle callback period.
+    Creates idle callback if it doesn't exist with a timeout of 1s.
+    */
+    public enqueueTask<T>(task: () => T){
```

Comment:
The ideal usage here - task is async (i.e. returns Promise<T>), and enqueueTask() returns me same type (Promise<T>) that gets completed when task has run (with results of that task).
I can build it as a separate layer (and make task return void), maybe it needs to be its own layer, but I expect 99% of usages to need this layer, so we may just build it into this layer.

Something like that:

const deferred = new Deferred<T>();
this.idleTaskList.push(() => {
   deferred.resolve(await task()); // await is optional here, but ideally, we know when task is over, before we run next task
});

return deferred.promise;